activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -27,23 +27,36 @@ def run
         end
 
         def records_by_owner
-          # owners can be duplicated when a relation has a collection association join
-          # #compare_by_identity makes such owners different hash keys
-          @records_by_owner ||= preloaded_records.each_with_object({}.compare_by_identity) do |record, result|
-            owners_by_key[convert_key(record[association_key_name])].each do |owner|
-              (result[owner] ||= []) << record
-            end
-          end
+          load_records unless defined?(@records_by_owner)
+
+          @records_by_owner
         end
 
         def preloaded_records
-          return @preloaded_records if defined?(@preloaded_records)
-          @preloaded_records = owner_keys.empty? ? [] : records_for(owner_keys)
+          load_records unless defined?(@preloaded_records)
+
+          @preloaded_records
         end
 
         private
           attr_reader :owners, :reflection, :preload_scope, :model, :klass
 
+          def load_records
+            # owners can be duplicated when a relation has a collection association join
+            # #compare_by_identity makes such owners different hash keys
+            @records_by_owner = {}.compare_by_identity
+            raw_records = owner_keys.empty? ? [] : records_for(owner_keys)
+
+            @preloaded_records = raw_records.select do |record|
+              assignments = owners_by_key[convert_key(record[association_key_name])].map do |owner|
+                entries = (@records_by_owner[owner] ||= [])
+                entries << record if reflection.collection? || entries.empty?
+              end
+
+              assignments.any?
```

Comment:
Let's use `!empty?` here

Reply:
Sure! I assume that this is in part to avoid the element tests `any?` performs?

I've applied this change (and expanded out the block to make this work), and avoided including `nil`s in the working array. This is a little more verbose than the `!assignments.compact.empty?` option -- either should work though.